src/packages/external-source/external-source.ts

```diff
@@ -0,0 +1,497 @@
+  if (keys.length === 1) {
+    const localSchemaCopy = structuredClone(baseExternalSourceSchema);
+
+    const eventTypeName = keys[0];
+    const eventTypeSchema = {
+      ...defs.event_types[eventTypeName],
+      additionalProperties: false,
+    };
+    const sourceTypeName = Object.keys(defs.source_type)[0];
+    const sourceTypeSchema = {
+      ...defs.source_type[sourceTypeName],
+      additionalProperties: false,
+    };
+
+    localSchemaCopy.properties.events.items.properties.attributes = eventTypeSchema;
+    localSchemaCopy.properties.events.items.properties.event_type_name = { const: eventTypeName };
+
+    // Insert def for "source" attributes
+    localSchemaCopy.properties.source.properties.attributes = sourceTypeSchema;
+
+    const localAjv = new Ajv();
+    return localAjv.compile(localSchemaCopy);
+  }
+
+  // Handle n event types
+  for (let i = 0; i < keys.length - 1; i++) {
+    const key = keys[i];
+    // Create tree of if/else/then statements to support validating different types
+    ifThenElsePointer['if'] = {
+      properties: {
+        event_type_name: {
+          const: key,
+        },
+      },
+    };
+    ifThenElsePointer['then'] = {
+      properties: {
+        attributes: {
+          $ref: `#/$defs/event_types/${key}`,
+        },
+      },
+    };
+    ifThenElsePointer['else'] = {};
+    ifThenElsePointer = ifThenElsePointer['else'];
+  }
+
+  const key = keys[keys.length - 1];
+  ifThenElsePointer['properties'] = {
+    attributes: {
+      $ref: `#/$defs/event_types/${key}`,
+    },
+  };
+
+  const localSchemaCopy = structuredClone(baseExternalSourceSchema);
+  localSchemaCopy.properties.events.items['if'] = ifThenElse['if'];
+  localSchemaCopy.properties.events.items['then'] = ifThenElse['then'];
+  localSchemaCopy.properties.events.items['else'] = ifThenElse['else'];
+
+  // Insert def for "source" attributes
+  const sourceTypeKey = Object.keys(defs.source_type)[0];
+  localSchemaCopy.properties.source.properties.attributes = { $ref: `#/$defs/source_type/${sourceTypeKey}` };
+
+  // Add defs
+  localSchemaCopy.$defs = {
+    event_types: {},
+    source_type: {
+      [sourceTypeKey]: {
+        ...defs.source_type[sourceTypeKey],
+        additionalProperties: false,
+      },
+    },
+  };
+  for (const eventType of keys) {
+    localSchemaCopy.$defs.event_types[eventType] = {
+      ...defs.event_types[eventType],
+      additionalProperties: false,
+    };
+  }
+
+  // Compile & return full schema with 'defs' added
+  const localAjv = new Ajv();
+  return localAjv.compile(localSchemaCopy);
+}
+
+async function uploadExternalSourceEventTypes(req: Request, res: Response) {
+  const authorizationHeader = req.get('authorization');
+
+  const {
+    headers: { 'x-hasura-role': roleHeader, 'x-hasura-user-id': userHeader },
+  } = req;
+
+  const { body } = req;
+  const { event_types, source_types } = body;
```

Comment:
This can be done at once:
```suggestion
  const {
    headers: { 'x-hasura-role': roleHeader, 'x-hasura-user-id': userHeader },
    body: {event_types, source_types}
  } = req;
```

This comment applies to multiple functions in this file

Reply:
Implemented!